examples/llm_plus_gnn/g_retriever.py

```diff
@@ -0,0 +1,472 @@
+            prep_time = round(time.time() - since, 2)
+            print("Total Prep Time (prep_time) =", prep_time)
+            print("Training beginning...")
+        epoch_str = f"Epoch: {epoch + 1}|{num_epochs}"
+        loader = tqdm(train_loader, desc=epoch_str)
+        for step, batch in enumerate(loader):
+            optimizer.zero_grad()
+            loss = loss_fn(model, batch, model_save_name)
+            loss.backward()
+
+            clip_grad_norm_(optimizer.param_groups[0]['params'], 0.1)
+
+            if (step + 1) % grad_steps == 0:
+                adjust_learning_rate(optimizer.param_groups[0], lr,
+                                     step / len(train_loader) + epoch)
+
+            optimizer.step()
+            epoch_loss = epoch_loss + loss.item()
+
+            if (step + 1) % grad_steps == 0:
+                lr = optimizer.param_groups[0]["lr"]
+        train_loss = epoch_loss / len(train_loader)
+        print(epoch_str + f",Train Loss (Epoch Mean): {train_loss}")
+
+        val_loss = 0.
+        eval_output = []
+        model.eval()
+        with torch.no_grad():
+            for step, batch in enumerate(val_loader):
+                loss = loss_fn(model, batch, model_save_name)
+                val_loss += loss.item()
+            val_loss = val_loss / len(val_loader)
+            print(epoch_str + f", Val Loss: {val_loss}")
+        if checkpointing and val_loss < best_val_loss:
+            print("Checkpointing best val loss model...")
+            best_val_loss = val_loss
+            best_epoch = epoch
+            save_params_dict(model, model_save_name + "_best_val_loss_ckpt.pt")
+    torch.cuda.empty_cache()
+    torch.cuda.reset_max_memory_allocated()
+
+    # Step 5 Evaluating
+    if checkpointing and best_epoch != num_epochs - 1:
+        print("Loading best checkpoint...")
+        model = load_params_dict(model,
+                                 model_save_name + "_best_val_loss_ckpt.pt")
+    model.eval()
+    eval_output = []
+    print("Final Evaluation...")
+    progress_bar_test = tqdm(range(len(test_loader)))
+    for step, batch in enumerate(test_loader):
+        with torch.no_grad():
+            pred = inference_fn(model, batch, model_save_name)
+            eval_data = {
+                "pred": pred,
+                "question": batch.question,
+                "desc": batch.desc,
+                "label": batch.label
+            }
+            eval_output.append(eval_data)
+        progress_bar_test.update(1)
+
+    # Step 6 Post-processing & compute metrics
+    acc = compute_accuracy(eval_output)
+    print(f'Test Acc {acc}')
+    # save model
+    print("Saving Model...")
+    save_params_dict(model, model_save_name + ".pt")
+    print("Saving eval output for downstream demo...")
+    torch.save(eval_output, model_save_name + "_eval_outs.pt")
+    print("Done!")
+    return prep_time, dataset, eval_output
+
+
+def minimal_demo(gnn_llm_eval_outs, dataset, lr, epochs, batch_size,
```

Comment:
Maybe we can put this to a separate file? We can create a folder named with `g-retriever` and have files such as train.py, utils.py and compare.py. WDYT?

Reply:
in general i prefer standalone examples since they give users of our container a smoother out of the box experience. open to other options if @rusty1s feels strongly about it